sympy/functions/elementary/tests/test_piecewise.py

```diff
@@ -550,6 +687,300 @@ def test_S_srepr_is_identity():
+    assert f.integrate((t, t0, T)) == Piecewise(
+        (-t0*x + x*Min(T, Max(t0, t1)), T > t0),
+        (0, True))
+
+def test_issue_10122():
+    assert solve(abs(x) + abs(x - 1) - 1 > 0, x
+        ) == Or(And(-oo < x, x < 0), And(S.One < x, x < oo))
+
+
+def test_issue_4313():
+    u = Piecewise((0, x <= 0), (1, x >= a), (x/a, True))
+    e = (u - u.subs(x, y))**2/(x - y)**2
+    M = Max(0, a)
+    assert integrate(e,  x).expand() == Piecewise(
+        (Piecewise(
+            (0, x <= 0),
+            (-y**2/(a**2*x - a**2*y) + x/a**2 - 2*y*log(-y)/a**2 +
+                2*y*log(x - y)/a**2 - y/a**2, x <= M),
+            (-y**2/(-a**2*y + a**2*M) + 1/(-y + M) -
+                1/(x - y) - 2*y*log(-y)/a**2 + 2*y*log(-y +
+                M)/a**2 - y/a**2 + M/a**2, True)),
+        ((a <= y) & (y <= 0)) | ((y <= 0) & (y > -oo))),
+        (Piecewise(
+            (-1/(x - y), x <= 0),
+            (-a**2/(a**2*x - a**2*y) + 2*a*y/(a**2*x - a**2*y) -
+                y**2/(a**2*x - a**2*y) + 2*log(-y)/a - 2*log(x - y)/a +
+                2/a + x/a**2 - 2*y*log(-y)/a**2 + 2*y*log(x - y)/a**2 -
+                y/a**2, x <= M),
+            (-a**2/(-a**2*y + a**2*M) + 2*a*y/(-a**2*y +
+                a**2*M) - y**2/(-a**2*y + a**2*M) +
+                2*log(-y)/a - 2*log(-y + M)/a + 2/a -
+                2*y*log(-y)/a**2 + 2*y*log(-y + M)/a**2 -
+                y/a**2 + M/a**2, True)),
+        a <= y),
+        (Piecewise(
+            (-y**2/(a**2*x - a**2*y), x <= 0),
+            (x/a**2 + y/a**2, x <= M),
+            (a**2/(-a**2*y + a**2*M) -
+                a**2/(a**2*x - a**2*y) - 2*a*y/(-a**2*y + a**2*M) +
+                2*a*y/(a**2*x - a**2*y) + y**2/(-a**2*y + a**2*M) -
+                y**2/(a**2*x - a**2*y) + y/a**2 + M/a**2, True)),
+        True))
+
+
+def test__intervals():
+    assert Piecewise((x + 2, Eq(x, 3)))._intervals(x) == [(3, 3, 5, 0)]
+    assert Piecewise(
+        (1, x > x + 1),
+        (Piecewise((1, x < x + 1)), 2*x < 2*x + 1),
+        (1, True))._intervals(x) == [(-oo, oo, 1, 1)]
+    assert Piecewise((1, Ne(x, I)), (0, True))._intervals(x) == [
+        (-oo, oo, 1, 0)]
+
+
+def test_containment():
+    a, b, c, d, e = [1, 2, 3, 4, 5]
+    p = (Piecewise((d, x > 1), (e, True))*
+        Piecewise((a, Abs(x - 1) < 1), (b, Abs(x - 2) < 2), (c, True)))
+    assert p.integrate(x).diff(x) == Piecewise(
+        (c*e, x <= 0),
+        (a*e, x <= 1),
+        (a*d, x < 2),  # this is what we want to get right
+        (b*d, x < 4),
+        (c*d, True))
+
+
+def test_piecewise_with_DiracDelta():
+    # XXX not sure what these should be
+    d1 = DiracDelta(x - 1)
+    assert integrate(d1, (x, -oo, oo)) == 1
+    assert integrate(d1, (x, 0, 2)) == 1
+    assert Piecewise((d1, Eq(x, 2)), (0, True)).integrate(x) == 0  # right
+    # XXX should the following be
+    # Piecewise((Heaviside(0), Eq(x, 1)), (1, x > 1), (0, True))?
+    assert Piecewise((d1, Eq(x, 1)), (0, True)).integrate(x) == 0
```

Comment:
I think this test should not be included. Although it is possible to say that `d1` is 0 for `x != 1`, its value at 1 cannot be considered as a number. There is no reasonable way of assigning some meaning to `Piecewise((d1, Eq(x, 1)), (0, True))`.

Reply:
From my (very limited) understanding of DD, I am not thinking there is a value I can assign other than DD(0) at x=1, but I was thinking of the answer in the inline comment based on the following:

```
d1=DiracDelta(x - 1)
p = Dummy(positive=True)
integrate(d1, (x, -oo, 1 - p))
0
integrate(d1, (x, -oo, 1))
Heaviside(0)
integrate(d1, (x, -oo, 1 +p))
1
```

Reply:
The expression `integrate(d1, (x, -oo, 1))` is also ambiguous since it is does not make clear whether 1 is included in the domain of integration or not. To define integrals of singular measures it is necessary to specify the domain exactly. The integral of `d1` over the open interval `(-oo, 1)` is 0 but over the half-closed interval `(-oo, 1]` it is 1. It should also be noted that the [fundamental theorem of calculus](https://en.wikipedia.org/wiki/Fundamental_theorem_of_calculus#Formal_statements) on the connection between integral and antiderivative is valid for continuous functions only. It cannot be used if the integrand is discontinuous at one of the limits of integration.

Reply:
I will delete the test and leave this as a new issue.